fiftyone/core/patches.py

```diff
@@ -0,0 +1,410 @@
+            this view
+        filtered_fields (None): a set of field names of list fields that are
+            filtered in this view
+    """
+
+    pass
+
+
+class EvaluationPatchView(_PatchView):
+    """A patch in an :class:`EvaluationPatchesView`.
+
+    :class:`EvaluationPatchView` instances should not be created manually; they
+    are generated by iterating over :class:`EvaluationPatchesView` instances.
+
+    Args:
+        doc: a :class:`fiftyone.core.odm.DatasetSampleDocument`
+        view: the :class:`EvaluationPatchesView` that the patch belongs to
+        selected_fields (None): a set of field names that this view is
+            restricted to
+        excluded_fields (None): a set of field names that are excluded from
+            this view
+        filtered_fields (None): a set of field names of list fields that are
+            filtered in this view
+    """
+
+    pass
+
+
+class _PatchesView(fov.DatasetView):
+    """Base class for :class:`fiftyone.core.view.DatasetView` classes that
+    contain sample patches.
+
+    Args:
+        source_collection: the
+            :class:`fiftyone.core.collections.SampleCollection` from which this
+            view was created
+        patches_stage: the :class:`fiftyone.core.stages.ViewStage` stage that
+            defines how the patches were extracted
+        patches_dataset: the :class:`fiftyone.core.dataset.Dataset` that serves
+            the patches in this view
+    """
+
+    def __init__(
+        self, source_collection, patches_stage, patches_dataset, _stages=None
+    ):
+        if _stages is None:
+            _stages = []
+
+        self._source_collection = source_collection
+        self._patches_stage = patches_stage
+        self._patches_dataset = patches_dataset
+        self.__stages = _stages
+
+    def __copy__(self):
+        return self.__class__(
+            self._source_collection,
+            deepcopy(self._patches_stage),
+            self._patches_dataset,
+            _stages=deepcopy(self.__stages),
+        )
+
+    @property
+    def _label_fields(self):
+        raise NotImplementedError("subclass must implement _label_fields")
+
+    @property
+    def _dataset(self):
+        return self._patches_dataset
+
+    @property
+    def _root_dataset(self):
+        return self._source_collection._root_dataset
+
+    @property
+    def _stages(self):
+        return self.__stages
+
+    @property
+    def _all_stages(self):
+        return (
+            self._source_collection.view()._all_stages
+            + [self._patches_stage]
+            + self.__stages
+        )
+
+    @property
+    def _element_str(self):
+        return "patch"
+
+    @property
+    def _elements_str(self):
+        return "patches"
+
+    @property
+    def name(self):
+        return self.dataset_name + "-patches"
+
+    def tag_samples(self, tags):
+        super().tag_samples(tags)
+
+        # Update source collection
+
+        def sync_fcn(view, fields=None):
```

Comment:
So tagging `samples` in an evaluation patches view tags the root dataset labels, and the patches document, but not the labels in the patches documents?

Reply:
Yeah you're correct in pointing out the inconsistency here. Anything that happens to the root dataset needs to be reflected in the "linked" labels in the patches/evaluation views.

I took a bit of a pause on finishing (and testing, btw) the tagging behavior to let the options bounce around in my head a bit to see if a good approach would come to mind...

It might be best if tagging patches at the patch level did *NOT* apply to the labels. In fact, since patches are purely a view concept, this would create data that is not propagated to the root dataset at all, similar to what happens if you were to, for example, add a new field to a patches view.

Another option we discussed last week would be that tagging a patch would apply that tag to the root sample, but the "reflection" property mentioned about would dictate that all other patches from the same sample would need to automatically inherit that tag as well, but I don't like the idea of tagging one patch in a view causing updates to other documents in the view...

Where are your thoughts on all this at the moment?

Reply:
This is confusing me

```py
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart")
patches = dataset.to_patches("predictions")
patches.tag_samples("test")

dataset.count_values("predictions.detections.tags")
patches.count_values("tags")
patches.count_values("predictions.tags")
dataset.to_patches("predictions").count_values("predictions.tags")
```

Out:
```
{'test': 5620}
{'test': 5620}
{}
{'test': 5620}
```

Reply:
I think the easiest and simplest conceptual path forward is just to change any relevant "SAMPLE" terminology to "PATCH" in the App when viewing patches views, and then remove the propagation of patch tag -> label tag stuff I started.

Reply:
Ok, I'll move forward with that. No propagation is better.

This is a tricky feature, so I have concerns in general.  I will say, it is easier for me to conceptualize patches as just labels the themselves rather than a type of sample.

Reply:
Okay I removed the propagation of sample tags back to the source dataset.

Feel free to voice any concerns you have over the behavior of this feature. Definitely want to make sure that it's as clear as possible to the user!

Reply:
It's definitely undesirable from an implementation standpoint that data is duplicated and new collections are created, but I couldn't figure out how to implement the evaluation and frames transformations in a single pipeline. And, if we were able to figure that out, I'm a bit concerned that the aggregations would be too slow to run repeatedly every time (eg) a new page of samples from the view is requested in the App.

Hopefully, if I've done my job well, this implementation detail is completely hidden from the App, though.

This implementation is *almost* completely hidden from users, but one unfortunate deviation from "regular" views is that updating the root dataset will not automatically affect a patches view; it must be re-instantiated in order to pull any changes.

But, then again, regular datasets and views are not completely "self-syncing", since, as Jason discovered last week, if you're working with a dataset in multiple shells, you would need to call `dataset_reload(hard=True)` in one shell if you happen to add a new sample field in another shell.

Along both of these lines, I was thinking that it may be time to add a `reload()` method to the public interface of datasets/views that will handle refreshing datasets and views to deal with the aforementioned issues.